src/main/java/org/scijava/ui/swing/widget/SwingFileListWidget.java

```diff
@@ -173,9 +208,29 @@ public void actionPerformed(final ActionEvent e) {
 			fileList = uiService.chooseFiles(null, fileList, filter, style);
 			if (fileList == null)
 				return;
-			for (File file : fileList) {
-				listModel.addElement(file);
+			fileList.forEach(file -> listModel.addElement(file));
+		} else if (e.getSource() == addFolderButton) {
+			final File[] result = new File[1];
+			try {
+				// NB: We show the JFileChooser on the EDT because otherwise there could
+				// be a deadlock, particularly on macOS. See scijava/scijava-ui-swing#28.
+				threadService.invoke(() -> {
```

Comment:
Instead of doing this: can we use `uiService.chooseFiles`? Then we can have this threading logic solved properly in a single place.

Reply:
Good suggestion, I'll change this.

Reply:
I ended up using `ui()` instead of `uiService` because I'd otherwise get the Legacy UI's directory chooser, which led to some of the Swing buttons not being rendered correctly (not "released" after drop), and the look-and-feel being changed when running the example script a second time.

When using `ui()`, we keep using the SwingUI for both the dialog and the file chooser, which worked fine in my tests.

@ctrueden if you have no concerns with this, I'd say this is ready to be merged.